### PR TITLE
LUCENE-9790: Hunspell: avoid slow dictionary lookup if the word's hash isn't there

### PR DESCRIPTION
**Pull request by donnerpeter**
# Description

FST traversal is the hottest spot in Hunspell, let's call it less frequently

# Solution

Add a hash-based Bloom filter before FST traversal to avoid it.

# Tests

None. `TestPerformance.de` became ~30% faster, `fr` and `en` — ~15% faster.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `master` branch.
- [x] I have run `./gradlew check`.
- [ ] I have added tests for my changes.
- [ ] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

### CODE REVIEW

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/Dictionary.java [hunk 1]
```diff
@@ -249,7 +254,9 @@ public void close() {
       readAffixFile(affixStream, decoder, flagEnumerator);
 
       // read dictionary entries
-      IndexOutput unsorted = mergeDictionaries(tempDir, tempFileNamePrefix, dictionaries, decoder);
+      IndexOutput unsorted = tempDir.createTempOutput(tempFileNamePrefix, "dat", IOContext.DEFAULT);
+      int wordCount = mergeDictionaries(dictionaries, decoder, unsorted);
```
Comment (donnerpeter) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/Dictionary.java:258]: Moved `unsorted` calculation outside of `mergeDictionaries` to be able to return `wordCount` instead

File: lucene/analysis/common/src/test/org/apache/lucene/analysis/hunspell/TestAllDictionaries.java [hunk 2]
```diff
@@ -160,7 +160,8 @@ public void testDictionariesLoadSuccessfully() throws Exception {
           try {
             Dictionary dic = loadDictionary(aff);
             totalMemory.addAndGet(RamUsageTester.sizeOf(dic));
-            totalWords.addAndGet(RamUsageTester.sizeOf(dic.words));
+            totalWords.addAndGet(
+                RamUsageTester.sizeOf(dic.words) + RamUsageTester.sizeOf(dic.wordHashes));
```
Comment (donnerpeter) [inline lucene/analysis/common/src/test/org/apache/lucene/analysis/hunspell/TestAllDictionaries.java:164]: About ~7% memory usage increase on average, at most 512KB

Comment (dweiss) [reply lucene/analysis/common/src/test/org/apache/lucene/analysis/hunspell/TestAllDictionaries.java:164]: These are actual numbers from the dictionaries you checked, correct? The code doesn't seem to have an upper limit (only highestOneBit(words*10)). I hope the dictionaries out there are reasonable (lexicon of spoken human languages certainly is bound, so I think it's a sane assumption).

Comment (donnerpeter) [reply lucene/analysis/common/src/test/org/apache/lucene/analysis/hunspell/TestAllDictionaries.java:164]: Yes, that's from wooorm and LibreOffice dictionaries

**Pull request merged**